### PR TITLE
cgroup: Add support to check systemd-internal cgroup limits against the kernel

### PR DESCRIPTION
**Pull request by cdown**
This patch series implements both a backend and a minimal `systemd-analyze dump` frontend for checking systemd-internal cgroup memory limits against kernfs.

Cc @poettering, since we discussed this last week.

---

We currently don't have any mitigations against another privileged user on the system messing with the cgroup hierarchy, bringing the system out of line with what we've set in systemd. We also don't have any real way to surface this to the user.

There are a few possible solutions:

1. Maintaining our own cgroup tree with the new fsopen API and having a read-only copy for everyone else. However, there are some complications on this front, and this may be infeasible in some environments. I'd rate this as an effort that's tangential to this patch.
2. Actively checking for changes with {fa,i}notify and changing them back afterwards to match our configuration again. This is also possible, but it's also good to have a way to do passive monitoring of the situation without taking hard action. Also, currently daemons like senpai do actually need to modify the tree behind systemd's back (although hopefully this should be more integrated soon).

This patch implements another option, where one can, on demand, monitor deviations in cgroup memory configuration from systemd's internal state. Currently the only consumer is `systemd-analyze dump`, but the interface is generic enough that it can also be exposed elsewhere later (for example, over D-Bus).

Currently only memory limit style properties are supported, but later I also plan to expand this out to other properties that systemd should have ultimate control over.

### CODE REVIEW

File: src/core/cgroup.c [hunk 1]
```diff
@@ -207,6 +207,109 @@ void cgroup_context_done(CGroupContext *c) {
...
 
+static int unit_get_kernel_memory_limit(Unit *u, const char *file, uint64_t *ret) {
+        _cleanup_free_ char *raw_kval = NULL;
+        uint64_t kval;
+        int r;
+
+        assert(u);
+
+        if (!u->cgroup_realized || UNIT_IS_INACTIVE_OR_FAILED(unit_active_state(u)))
+                return -EOPNOTSUPP;
```
Comment (keszybz) [inline src/core/cgroup.c]: Hmm, @poettering likes when EOPNOTSUPP is only used for stuff that is actually unsupported. Some different errno should be used, maybe EOWNERDEAD?

File: src/core/cgroup.c [hunk 2]
```diff
@@ -207,6 +207,109 @@ void cgroup_context_done(CGroupContext *c) {
...
+
+        if (r < 0)
+                return log_debug_errno(r, "Failed to determine cgroup hierarchy version: %m");
+
+        /* Unsupported on v1.
+         *
+         * We don't return ENOENT, since that could actually mask a genuine problem where somebody else has
+         * silently masked the controller. */
+        if (r == 0)
+                return -EOPNOTSUPP;
```
Comment (keszybz) [inline src/core/cgroup.c]: Same here, maybe -ENODATA?

File: src/core/cgroup.c [hunk 3]
```diff
@@ -207,6 +207,109 @@ void cgroup_context_done(CGroupContext *c) {
...
+
+        /* The root slice doesn't have any controller files, so we can't compare anything. */
+        if (unit_has_name(u, SPECIAL_ROOT_SLICE))
+                return -EOPNOTSUPP;
```
Comment (keszybz) [inline src/core/cgroup.c]: -ENODATA.

File: src/core/cgroup.c [hunk 4]
```diff
@@ -207,6 +207,109 @@ void cgroup_context_done(CGroupContext *c) {
...
+
+        /* Even if you write a particular number of bytes into a cgroup memory file, it always returns that
+         * number page-aligned (since the kernel internally stores cgroup limits in pages). As such, so long
+         * as it aligns properly, everything is cricket. */
+        if (*sval != CGROUP_LIMIT_MAX)
+                *sval = (uint64_t)PAGE_ALIGN(*sval);
+
+        r = unit_get_kernel_memory_limit(u, file, kval);
+        if (r < 0)
+            return r;
```
Comment (keszybz) [inline src/core/cgroup.c]: Indentation.

File: src/core/cgroup.c [hunk 5]
```diff
@@ -207,6 +207,109 @@ void cgroup_context_done(CGroupContext *c) {
...
+        int r;
+
+        /* Compare kernel memcg configuration against our internal systemd state. Unsupported (and will
+         * return -EOPNOTSUPP) on cgroup v1.
+         *
+         * Returns:
+         *
+         * <0: On error.
+         *  0: If the kernel memory setting doesn't match our configuration.
+         * >0: If the kernel memory setting matches our configuration. */
```
Comment (keszybz) [inline src/core/cgroup.c]: Maybe describe in the comment also what `sval` and `kval` return. Maybe also rename them to `ret_unit_value` and `ret_kernel_value` for clarity.

File: src/core/cgroup.c [hunk 6]
```diff
@@ -207,6 +207,109 @@ void cgroup_context_done(CGroupContext *c) {
...
+                *sval = c->memory_swap_max;
+                file = "memory.swap.max";
+        } else
+                return -EINVAL;
+
+        /* Even if you write a particular number of bytes into a cgroup memory file, it always returns that
+         * number page-aligned (since the kernel internally stores cgroup limits in pages). As such, so long
+         * as it aligns properly, everything is cricket. */
+        if (*sval != CGROUP_LIMIT_MAX)
+                *sval = (uint64_t)PAGE_ALIGN(*sval);
```
Comment (keszybz) [inline src/core/cgroup.c]: Not that it matter too much in this case, but we generally prefer to atomatically either update all output values on sucess, or none at all on failure. So it'd be nicer to use an internal variable for this, and only assign the output variable once we know that no more operations can fail.

File: src/core/cgroup.c [hunk 7]
```diff
@@ -336,36 +336,36 @@ void cgroup_context_dump(CGroupContext *c, FILE* f, const char *prefix) {
...
-                "%sMemoryLow=%" PRIu64 "\n"
-                "%sMemoryHigh=%" PRIu64 "\n"
-                "%sMemoryMax=%" PRIu64 "\n"
-                "%sMemorySwapMax=%" PRIu64 "\n"
-                "%sMemoryLimit=%" PRIu64 "\n"
-                "%sTasksMax=%" PRIu64 "\n"
-                "%sDevicePolicy=%s\n"
-                "%sDisableControllers=%s\n"
-                "%sDelegate=%s\n",
+                "%sCPUAccounting: %s\n"
```
Comment (keszybz) [inline src/core/cgroup.c:398]: Indeed. I wonder how I never noticed that.

File: src/core/cgroup.c [hunk 8]
```diff
@@ -214,12 +214,12 @@ static int unit_get_kernel_memory_limit(Unit *u, const char *file, uint64_t *ret
...
         assert(u);
 
-        if (!u->cgroup_realized || UNIT_IS_INACTIVE_OR_FAILED(unit_active_state(u)))
+        if (!u->cgroup_realized)
                 return -EOPNOTSUPP;
 
         r = cg_get_attribute("memory", u->cgroup_path, file, &raw_kval);
         if (r < 0)
-                return log_debug_errno(r, "Failed to get value of %s from %s: %m", file, u->cgroup_path);
+                return r;
```
Comment (keszybz) [inline src/core/cgroup.c:223]: Shouldn't this chunk be moved to the first patch?

File: src/core/cgroup.c [hunk 9]
```diff
@@ -305,12 +305,43 @@ static int unit_compare_memory_limit(Unit *u, const char *property_name, uint64_
...
         r = unit_get_kernel_memory_limit(u, file, kval);
         if (r < 0)
-            return r;
+                return log_unit_debug_errno(u, r, "Failed to parse %s: %m", file);
 
         return *kval == *sval;
 }
 
-void cgroup_context_dump(CGroupContext *c, FILE* f, const char *prefix) {
+#define FORMAT_CGROUP_DIFF_MAX 128
```
Comment (keszybz) [inline src/core/cgroup.c:331]: If there's an error, 128 characters will not be enough, " (error getting kernel value: %s)" leaves only 30 characters for that... OTOH, making this dynamic seems overkill. Maybe just make this 1024?

**Comment by keszybz** [comment]
"kernfs" is not a thing really. Maybe just say "kernel" to avoid confusion?

File: src/core/cgroup.c (hunk 8 above)
Comment (cdown) [reply src/core/cgroup.c:223]: Oh yeah, that's a mistake. Thanks.

File: src/core/cgroup.c (hunk 9 above)
Comment (cdown) [reply src/core/cgroup.c:331]: Hmm? After `(error getting kernel value: %s)` we still have 96 characters to go to store the actual error, no?

No objection to just doing 1024 though

Comment (keszybz) [reply src/core/cgroup.c:331]: Oh, right. 31 is the length of the static string... The longest string strerror currently returns is 49. So yeah, 128 is enough, please ignore my comment.

**Comment by cdown** [comment]
Thanks for the thorough review! I believe all comments are addressed now. :-)

File: src/core/cgroup.c [hunk 10]
```diff
@@ -208,6 +208,117 @@ void cgroup_context_done(CGroupContext *c) {
...
+         *
+         * The following values are only guaranteed to be populated on return >=0:
+         *
+         * - ret_unit_value will contain our internal expected value for the unit, page-aligned.
+         * - ret_kernel_value will contain the actual value presented by the kernel. */
+
+        assert(u);
+
+        r = cg_all_unified();
+
```
Comment (keszybz) [inline src/core/cgroup.c]: Spurious newline.

**Comment by cdown** [comment]
2 changes after accept:

1. Turns out cgroup v2 docs are wrong and page alignment is always down. I'll go fix that in kernel. For now I created a PAGE_ALIGN_DOWN macro and use it instead of PAGE_ALIGN.
2. Page alignment for these files is pretty useless anyway, so I'll probably change them to not be page aligned. Added the code to handle future kernels with my patch in.

**Comment by keszybz** [comment]
> 2 changes after accept:
> 
>     1. Turns out cgroup v2 docs are wrong and page alignment is always down. I'll go fix that in kernel docs. For now I created a PAGE_ALIGN_DOWN macro and use it instead of PAGE_ALIGN.

Oh, I noticed the downward alignment when testing this, and I had a thought "oh, this is strange", but I left it at that ;)

>     2. Page alignment for these files is pretty useless anyway, so I'll probably change them to not be page aligned. Added the code to handle future kernels with my patch in.

Right.

**Pull request merged**